csharp/ql/src/semmle/code/csharp/PrintAst.qll

```diff
@@ -120,6 +120,16 @@ private predicate isInterestingBaseType(ValueOrRefType base) {
  */
 private newtype TPrintAstNode =
   TElementNode(Element element) { shouldPrint(element, _) } or
+  TTopLevelTypeMentionNode(TypeMention typeMention) {
+    shouldPrint(typeMention.getTarget(), _) and
+    not isNotNeeded(typeMention.getTarget().getParent*()) and
+    not typeMention.getTarget().getParent*() instanceof TypeParameter
+  } or
+  TNestedTypeMentionNode(TypeMention typeMention) {
+    shouldPrint(typeMention.getParent+().getTarget(), _) and
+    not isNotNeeded(typeMention.getParent+().getTarget().getParent*()) and
+    not typeMention.getParent+().getTarget().getParent*() instanceof TypeParameter
+  } or
```

Comment:
Is there a good reason to distinguish nested and top-level type mentions? It seems that something like this would encompass both cases:
```ql
TTypeMentionNode(TypeMention typeMention) {
  exists(Element target |
    target = typeMention.getParent*().getTarget() and
    shouldPrint(target, _) and
    not isNotNeeded(target.getParent*()) and
    not target.getParent*() instanceof TypeParameter
  )
}
```

Reply:
We need to differentiate between the two, because we only want to have top level type mentions directly under `ElementNode`s and `BaseTypesNode`s.

Reply:
But that could be handled by restricting to those `TypeMentionNode`s for which `getTarget()` exists.

Reply:
Makes sense. Let me adjust it.